### ISSUE TITLE
method.truncate_track not overridden and lacks ACT check

### ISSUE THREAD

**Issue by LeeReidCSIRO** [MRtrix3/mrtrix3#1234]
In advance: please excuse me if this is a wild goose chase – I don’t have any ACT-ready data to test this on. These also aren't issues that affect me, but they might affect others now, or be a headache for you in the future.

I’ve just been playing with the MRtrix source code, particularly with relevance to `method.truncate_track(tck, max_size_at_backtrack, revert_step);`

I needed this method for a mod I’m working on and found what I can only assume (?) are two bugs when combined with iFOD2:

1)	iFOD2 has its own version of this method but this lacks the override keyword. Thus, when you call `method.truncate_track` using `iFOD2`, it calls the version in `MethodBase`
Whether this currently impacts ACT, I’m not sure. Easy enough to test (Add `WARN("not a problem!");` in `iFOD2::method.truncate_track` and run ACT with backtracking) but – as said previously, I don’t have any 5TT etc files ready to investigate.
The solution to this is obviously to implement `virtual…override` in the relevant classes.
2)	Within the `iFOD.truncate_track` method, the final line:
`act().sgm_depth = (act().sgm_depth > points_to_remove) ? act().sgm_depth - points_to_remove : 0;`
needs an additional check as to whether ACT is actually in progress. Currently if this is called when ACT is not in progress, the method throws an exception 
```
if(S.is_act()){
act().sgm_depth = (act().sgm_depth > points_to_remove) ? act().sgm_depth - points_to_remove : 0;
}
```

Lee

**Comment by Lestropie** [comment]
> 1. iFOD2 has its own version of this method but this lacks the override keyword

[`override`](http://en.cppreference.com/w/cpp/language/override) only indicates that a virtual function *must* be overriding an underlying virtual base function; i.e. it will cause a compiler error if the `override` keyword is used but no corresponding virtual function within a base class is found. This is handy for catching minor differences between virtual functions, and is good coding practise, but is purely optional in C++11, and isn't strictly required in order for a virtual function to be overloaded.

The sneakier part is that `truncate_track()` isn't actually virtual. In `cmd/tckgen.cpp`, you can see that rather than allocating memory for the tracking method and assigning it to a `MethodBase` pointer, we instead invoke `Exec<>::run()` using the algorithm method as the template parameter. This avoids the computational overhead of invoking virtual functions inside tight iteration loops, at the expense of longer compilation / larger executable file. Having said that, `MethodBase::reverse_track()` *shouldn't* be defined as virtual for the same reason, though it currently is; I'll fix that.

> Within the iFOD2 `truncate_track()` method, the final line ... needs an additional check as to whether ACT is actually in progress.

There's no mechanism by which `truncate_track()` can possibly be called unless ACT is being used; hence that line can't be executed unless ACT is being used. Having said that, I do test it exactly as you describe in the `MethodBase::truncate_track()` version...

The preference from a performance standpoint would be instead to add:
```
assert (act_method_additions)
```
in there, since that disappears in release mode compilation. However if you're making source code changes that require the use of this function in the absence of ACT, this assumption won't apply to you, and you'll need to make changes accordingly.

Keep those hawk eyes open! :+1:
Rob

**Comment by LeeReidCSIRO** [comment]
Thanks Rob

1) Pretty interesting. I've had a play and you're right - it will execute the correct one from _within exec_. In my code, the `exec.method` object was passed to another class that called `truncate_track`, which resulted in the base class (wrong) method being called. This resulted in the position of the track being rewound, but a selection of fields in the method object not being rewound/recalculated; ongoing behaviour of the class was then rather unstable... For someone from a non C++ background (me!), that rings of a design flaw (i.e. the class was rendered unstable by polymorphism because it replaced rather than overrode a method that poorly affected its internal state), but it's quite possible I'm lacking understanding here due to a lack of C++ knowledge and associated expectations. Sorry for taking up your time.

2) An assert or if() statement both seem sensible to me - the `if` statement just seems more useful as it makes the function stable in circumstances beyond its present use. Encapsulation, assumptions, yada yada... I'm fairly sceptical that this would introduce any real overhead - it's the only relevant spot in the program I've found that lacks an `if(S.is_act())`; one more can't hurt...

Anyway, for my purposes I'll be marching ahead with my fixes. You are free to ignore etc as you see fit. I just wanted to let you know what I found in case it helped prevent some headaches for you in the future :)

All the best,
Lee

**Comment by Lestropie** [comment]
> In my code, the `exec.method` object was passed to another class that called `truncate_track`, which resulted in the base class (wrong) method being called.

> For someone from a non C++ background (me!), that rings of a design flaw

Yes, you're entirely correct: The code was simply never vetted with the intention of being re-used in the kind of way that it sounds like you're doing. For the sake of enabling such use, I think that if the code is written in such a way as to properly support polymorphism, but polymorphism is never actually *invoked* by `tckgen` (i.e. `tckgen.cpp` still invokes each algorithm as a template parameter), then we shouldn't get any performance penalty; i.e. the `vtable` will be constructed but never read at run-time, the function location will still be hard-coded at compile-time, and the code will behave as expected in your context.

> The if statement just seems more useful as it makes the function stable in circumstances beyond its present use. 

> I'm fairly sceptical that this would introduce any real overhead

Again, if the code is being re-used elsewhere such that our assumptions / limited testing are no longer comprehensive, I'm quite happy to make fixes to enable such use. If anything it's awesome to hear that such re-use is actually taking place. We definitely put effort into making classes modular for use across our own code base, but we don't necessarily perform extensive code analysis for every class for every conceivable external use; it would just be too much non-constructive overhead. Hence why dialogue such as this is very much welcome :+1: 

We tend to be quite sensitive around performance issues, particularly when it comes to `tckgen`, because of the extent of the optimisations that have already taken place in order to get the speed to where it currently is. We try to avoid branching statements wherever possible in such circumstances because of the detrimental effect on processor pipelining; hence the `assert()` suggestion. But for a function like `truncate_track()` that is not called at every streamline step, performance is less of an issue, so one more `if` statement will indeed have little effect.

**Comment by Lestropie** [comment]
Closed by #1235.
